### PR TITLE
Various Fixes

### PR DESCRIPTION
**Pull request by timoxley**
Found some suspicious looking code while hunting for source of connection bug.

* Fix some flakey async behaviour in `storage-config.test`. Seemed the event possibly fired before the listener was attached.
* ~~Actually compare `nodeIds` & `subscribeNodeIds` when determining whether tracker instruction fulfilled. Currently it just checks the number of nodes are the same, rather than the actual node ids.~~ Not needed.
* Nullify connection event handlers if connection is no longer current. Ensures old connections can't interfere with new state.
* Resolve subsequent calls to `connect` with a promise that resolves the same way the initial call does.
* Increase default `newConnectionTimeout` from 5000 to 15000. @juslesan mentioned connections can sometimes take a long time to form.
* Delete connection immediately when instructing it to close.
* Early return in propagateMessage when no subscribers to avoid dangling else. 
* close/clear endpoint connections/queues after resetting everything else.

### CODE REVIEW

File: src/logic/Node.ts [hunk 1]
```diff
@@ -111,7 +111,7 @@ export class Node extends EventEmitter {
         this.bufferTimeoutInMs = opts.bufferTimeoutInMs || 60 * 1000
         this.bufferMaxSize = opts.bufferMaxSize || 10000
         this.disconnectionWaitTime = opts.disconnectionWaitTime || 30 * 1000
-        this.nodeConnectTimeout = opts.nodeConnectTimeout || 4000
+        this.nodeConnectTimeout = opts.nodeConnectTimeout || 15000
```
Comment (juslesan) [inline src/logic/Node.ts:114]: This is likely a necessary increase based on the AWS experiments.
https://github.com/streamr-dev/network/pull/544#discussion_r616458096

File: src/connection/Connection.ts [hunk 2]
```diff
@@ -60,7 +60,7 @@ export class Connection {
     private connection: PeerConnection | null
     private dataChannel: DataChannel | null
     private paused: boolean
-    private lastState: string | null
+    public lastState: string | null
```
Comment (harbu) [inline src/connection/Connection.ts]: Keeping this private and providing a getter to fetch its value provides more guarantees about the value being only modified within this class, which helps the reader of this class reason about where and when this value is modified.

File: test/integration/storage-config.test.ts [hunk 3]
```diff
@@ -64,13 +64,17 @@ describe('storage node', () => {
             storages: [storage as Storage],
             storageConfig: config
         })
+
+        // @ts-expect-error private field
+        const t1 = waitForEvent(tracker.trackerServer, TrackerServerEvent.NODE_STATUS_RECEIVED)
```
Comment (harbu) [inline test/integration/storage-config.test.ts:69]: Did this alleviate the flakyness then?

File: src/connection/WebRtcEndpoint.ts [hunk 4]
```diff
@@ -147,7 +147,7 @@ export class WebRtcEndpoint extends EventEmitter {
             })
     }
 
-    connect(
+    async connect(
```
Comment (harbu) [inline src/connection/WebRtcEndpoint.ts:154]: why `async`, `await` not used?

File: src/connection/WebRtcEndpoint.ts [hunk 5]
```diff
@@ -175,17 +187,29 @@ export class WebRtcEndpoint extends EventEmitter {
             newConnectionTimeout: this.newConnectionTimeout,
             pingInterval: this.pingInterval,
             onLocalDescription: (type, description) => {
+                if (this.connections[targetPeerId] !== connection) {
```
Comment (harbu) [inline src/connection/WebRtcEndpoint.ts]: What is the relationship with the guards here and the ones you wrote here https://github.com/streamr-dev/network/pull/544/files#diff-4e51e534b6625d3469234524fa87f3e6c107ffe5c460926d93b001e939d3561dR140-R180. Are all these guards needed?

File: src/logic/Node.ts [hunk 6]
```diff
@@ -320,9 +320,12 @@ export class Node extends EventEmitter {
         this.logger.debug('subscribed to %j and unsubscribed from %j (streamId=%s, counter=%d)',
             subscribeNodeIds, unsubscribeNodeIds, streamId, counter)
 
-        if (subscribeNodeIds.length !== nodeIds.length) {
+        if (JSON.stringify(subscribeNodeIds.slice().sort()) !== JSON.stringify(nodeIds.slice().sort())) {
```
Comment (harbu) [inline src/logic/Node.ts]: Why would the node ids be different though? The values of `subscribeNodeIds`  will always be a subset of `nodeIds` (unless there is a bug elsewhere in the code)?

File: src/logic/Node.ts [hunk 7]
```diff
@@ -383,52 +386,57 @@ export class Node extends EventEmitter {
...
+        if (!subscribers.length) {
+            this.logger.warn('put back to buffer because could not propagate to %d nodes or more: %j',
+                MIN_NUM_OF_OUTBOUND_NODES_FOR_PROPAGATION, streamMessage.messageId)
+            this.logger.debug('streams: %j', this.streams)
             this.seenButNotPropagatedSet.add(streamMessage)
             this.messageBuffer.put(streamIdAndPartition.key(), [streamMessage, source])
+            return
         }
+
+        this.logger.info('propagating to subscribers: %s', subscribers.join(', '))
```
Comment (harbu) [inline src/logic/Node.ts]: Again, wondering if `info` should be here the default w.r.t. production logging needs

File: src/logic/Node.ts [hunk 8]
```diff
@@ -383,52 +386,57 @@ export class Node extends EventEmitter {
...
+        subscribers.forEach(async (subscriber) => {
+            try {
+                this.logger.info('propagating to %s >', subscriber)
```
Comment (harbu) [inline src/logic/Node.ts]: same on logging levels

File: src/connection/Connection.ts (hunk 2 above)
Comment (timoxley) [reply src/connection/Connection.ts]: This seems awfully Java but fair point, if it's private it can't possibly be mutated externally.

File: test/integration/storage-config.test.ts (hunk 3 above)
Comment (timoxley) [reply test/integration/storage-config.test.ts:69]: Yes, it appeared to, though maybe I just got lucky 3x in a row? There's actually a bunch of similar potential issues in the tests, one solution would be to wait for event loop to turn before starting anything that could trigger an event so callers always have a chance to attach listeners.

File: src/connection/WebRtcEndpoint.ts (hunk 4 above)
Comment (timoxley) [reply src/connection/WebRtcEndpoint.ts:154]: Any time you're returning a promise you should always use `async`, since this effectively wraps the whole function in a `try { … } catch(err) { return Promise.reject(err) }`. 

Otherwise you can end up with accidental uncaught sync errors e.g.

```js
function setup() {
  throw new Error('oops') // unintentional error
}

function connect() {
  setup()
  return new Promise((resolve) => resolve())
}

connect().catch(console.log) // won't catch because setup() threw synchronously
```

Should we crash for unintentional sync exceptions? Perhaps, but the issue is that it's very easy to accidentally create sync errors e.g. add some argument validation in `setup()` or something.

For the least surprises, when a function returns promises i.e. is async, it should only ever return promises, errors should always become rejections i.e. [Do not release Zalgo](https://blog.izs.me/2013/08/designing-apis-for-asynchrony). 

Adding `async` is low cost, but adds a *guarantee* that the function will not throw synchronously, ever, and the writer nor consumer will never have to think about whether that case is even possible.

File: src/connection/WebRtcEndpoint.ts (hunk 5 above)
Comment (timoxley) [reply src/connection/WebRtcEndpoint.ts]: Yeah perhaps not all needed, though I didn't see anything that made it obvious that it was 100% impossible for either:

1. a connection to be finished and not (yet) removed from list of connections
2. a connection to be removed from list of connections and not (yet) finished

I'll take another look into cleaning these up without need to keep checking the guards in all cases, agreed it's error prone and a smell.

File: src/logic/Node.ts (hunk 6 above)
Comment (timoxley) [reply src/logic/Node.ts]: ahh yes I see, you're right. Maybe just needs a comment, looked like a weird assumption but now I see it makes sense.

File: src/logic/Node.ts (hunk 7 above)
Comment (timoxley) [reply src/logic/Node.ts]: Will reduce to `debug`

File: src/logic/Node.ts (hunk 8 above)
Comment (timoxley) [reply src/logic/Node.ts]: Removed these logs entirely.

File: src/connection/WebRtcEndpoint.ts (hunk 5 above)
Comment (timoxley) [reply src/connection/WebRtcEndpoint.ts]: Converted `Connection` into an `EventEmitter` in 608fbc71

Same deal as the previous change to wrap `PeerConnection`/`DataChannel` event handling in an `EventEmitter`, makes it trivial to turn off all listeners immediately with `removeAllListeners`.

File: src/connection/Connection.ts (hunk 2 above)
Comment (timoxley) [reply src/connection/Connection.ts]: Done.

**Comment by timoxley** [comment]
Updated, needs re-review, in particular note the connection timeout gets reset when state moves to `connecting`, issue being I was seeing multiple second delays between connection state `null` (creation) and `connecting`:
https://github.com/streamr-dev/network/pull/546/commits/5c1afa45642651bfed213e8ec4179d0cc8c3a292

Not sure why, there was no logging from that side during the delay but might be that it needs to wait for other end to time out before it can try again, so it always needs 2x timeout? Not sure.

File: src/connection/WebRtcEndpoint.ts (hunk 4 above)
Comment (harbu) [reply src/connection/WebRtcEndpoint.ts:154]: I haven't thought about it from this angle before, but it makes sense to me now. Essentially the function's interface becomes easier to work with, since instead of having to

- (a) expect a sync error to be thrown
- (b) expect a rejection error
- (c) expect resolved value

the first case goes away, and you are left as the user of the function to handle only (b) and (c). Did I understand correctly @timoxley ?

File: src/connection/Connection.ts [hunk 9]
```diff
@@ -73,7 +66,26 @@ function DataChannelEmitter(dataChannel: DataChannel) {
...
+    message: (msg: string)  => void
+    close: (err?: Error) => void
+    error: (err: Error) => void
+    bufferLow: () => void
+    bufferHigh: () => void
+}
+
+// reminder: only use Connection emitter for external handlers
+// to make it safe for consumers to call removeAllListeners
+// i.e. no this.on('event')
```
Comment (harbu) [inline src/connection/Connection.ts:102]: :+1:

File: src/connection/Connection.ts [hunk 10]
```diff
@@ -73,7 +66,26 @@ function DataChannelEmitter(dataChannel: DataChannel) {
...
+export const ConnectionEmitter = EventEmitter as { new(): StrictEventEmitter<EventEmitter, Events> }
```
Comment (harbu) [inline src/connection/Connection.ts:103]: Having trouble wrapping my head around this type definition.

Am I understanding this correctly: you are using `EventEmitter`  as the base type but then extending it so that its constructor will result in objects of type `StrictEventEmitter<EventEmitter, Events>`?

File: src/connection/Connection.ts [hunk 11]
```diff
@@ -428,6 +385,45 @@ export class Connection {
         }
     }
 
+    private onStateChange(state: string): void {
```
Comment (harbu) [inline src/connection/Connection.ts:403]: Ah looks like this and the other PR has some duplication. Looks like the methods here are private which I was asking about on the other PR.

File: src/connection/WebRtcEndpoint.ts [hunk 12]
```diff
@@ -99,7 +99,7 @@ export class WebRtcEndpoint extends EventEmitter {
                 connection.setPeerInfo(PeerInfo.fromObject(originatorInfo))
                 connection.setRemoteDescription(description, 'answer' as DescriptionType.Answer)
             } else {
-                this.logger.warn('unexpected rtcAnswer from %s: %s', originatorInfo, description)
+                this.logger.warn('unexpected rtcAnswer from %s: %s', peerId, description)
```
Comment (harbu) [inline src/connection/WebRtcEndpoint.ts:102]: I have thought it better to move towards printing `peerInfo` details in logs instead of just ids because the ids are Ethereum addresses which makes it hard to follow what is going on. PeerInfo#toString will print the peer name as well as the first characters of the address: https://github.com/streamr-dev/network/blob/af6d1c571ce27f5b3ce741e4876d5a45df76f6c0/src/connection/PeerInfo.ts#L82-L84

Comment (harbu) [reply src/connection/WebRtcEndpoint.ts:102]: In the tests of network, ids are set artificially to be any string but in the broker-end they are always Ethereum addresses.

File: src/connection/WebRtcEndpoint.ts [hunk 13]
```diff
@@ -109,21 +109,25 @@ export class WebRtcEndpoint extends EventEmitter {
...
             this.connect(peerId, routerId, isOffering).catch((err) => {
-                this.logger.warn('connectListener induced connection failed, reason %s', err)
+                this.logger.warn('connectListener induced connection from %s failed, reason %s', peerId, err)
             })
         })
 
         rtcSignaller.setErrorListener(({ targetNode, errorCode }: ErrorOptions) => {
             const error = new WebRtcError(`RTC error ${errorCode} while attempting to signal with ${targetNode}`)
-            this.emit(`errored:${targetNode}`, error)
+            const connection = this.connections[targetNode]
```
Comment (harbu) [inline src/connection/WebRtcEndpoint.ts:126]: Hmm.. this change has some consequences, but I'm guessing it will be more on the side of fixing issues.

**Review by harbu — APPROVED** [review]

**Review by juslesan — APPROVED** [review]

File: src/connection/Connection.ts (hunk 10 above)
Comment (timoxley) [reply src/connection/Connection.ts:103]: Sort of, more that it's `extending EventEmitter` but overriding (not extending) the type so it defines a constructor (`new`) that gives you back an `EventEmitter` extended by `StrictEventEmitter`.

https://github.com/bterlson/strict-event-emitter-types#usage-with-subclasses

File: src/connection/WebRtcEndpoint.ts (hunk 12 above)
Comment (timoxley) [reply src/connection/WebRtcEndpoint.ts:102]: That's a good idea.

File: src/connection/Connection.ts (hunk 10 above)
Comment (timoxley) [reply src/connection/Connection.ts:103]: Related: https://github.com/streamr-dev/network/pull/544#discussion_r619080467

File: src/connection/WebRtcEndpoint.ts (hunk 12 above)
Comment (timoxley) [reply src/connection/WebRtcEndpoint.ts:102]: Ideally we'd use `peerInfo` everywhere instead of strings, but currently many places only have one of the name/id or the address, but not both. 
I tried changing `Connection` to take `PeerInfo` objects instead of strings for `selfId` and `targetPeerId`, and was able to make this change work pretty straightforwardly.

But then things got tricky once I was in `handleTrackerInstruction` calling `nodeToNode.connectToNode` in `logic/Node.ts`, as it wasn't immediately clear how to get the bits needed to build a `PeerInfo`, so I just dropped it.

The logging changes here are mostly about fixing `from: [object Object]` from appearing in the logs, let's land this as-is.

File: src/connection/WebRtcEndpoint.ts (hunk 4 above)
Comment (timoxley) [reply src/connection/WebRtcEndpoint.ts:154]: 100% correct 👍
